File: Zig/disjoint_set_utils/src/disjoint_set.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

/// Disjoint Set Union (Union-Find) data structure
/// Implements efficient union and find operations with:
/// - Path compression in find()
/// - Union by rank optimization
/// - O(α(n)) amortized time complexity (inverse Ackermann function)
pub fn DisjointSet(comptime T: type) type {
    return struct {
        const Self = @This();

        /// Node structure for each element
        const Node = struct {
            parent: usize,
            rank: usize,
        };

        nodes: []Node,
        values: []T,
        count: usize,
        allocator: Allocator,

        /// Initialize a new DisjointSet with given elements
        pub fn init(allocator: Allocator, elements: []const T) Allocator.Error!Self {
            const nodes = try allocator.alloc(Node, elements.len);
            errdefer allocator.free(nodes);

            const values = try allocator.alloc(T, elements.len);
            errdefer allocator.free(values);

            @memcpy(values, elements);

            // Initialize each element as its own parent (separate set)
            for (nodes, 0..) |*node, i| {
                node.parent = i;
                node.rank = 0;
            }

            return .{
                .nodes = nodes,
                .values = values,
                .count = elements.len,
                .allocator = allocator,
            };
        }

        /// Initialize with a range of indices [0, num_elements)
        pub fn initRange(allocator: Allocator, num_elements: usize) Allocator.Error!Self {
            const nodes = try allocator.alloc(Node, num_elements);
            errdefer allocator.free(nodes);

            const values = try allocator.alloc(T, num_elements);
            errdefer allocator.free(values);

            // Initialize indices as values
            for (nodes, values, 0..) |*node, *value, i| {
                node.parent = i;
                node.rank = 0;
                value.* = @intCast(i);
            }

            return .{
                .nodes = nodes,
                .values = values,
                .count = num_elements,
                .allocator = allocator,
            };
        }

        /// Free all allocated memory
        pub fn deinit(self: *Self) void {
            self.allocator.free(self.nodes);
            self.allocator.free(self.values);
            self.* = undefined;
        }

        /// Find the root/representative of the set containing element at index
        /// Uses path compression for optimization
        pub fn find(self: *Self, index: usize) usize {
            std.debug.assert(index < self.count);

            // Path compression: make every node on the path point directly to root
            if (self.nodes[index].parent != index) {
                self.nodes[index].parent = self.find(self.nodes[index].parent);
            }
            return self.nodes[index].parent;
        }

        /// Find root without path compression (const version)
        pub fn findConst(self: Self, index: usize) usize {
            std.debug.assert(index < self.count);

            var current = index;
            while (self.nodes[current].parent != current) {
                current = self.nodes[current].parent;
            }
            return current;
        }

        /// Check if two elements are in the same set
        pub fn connected(self: *Self, a: usize, b: usize) bool {
            return self.find(a) == self.find(b);
        }

        /// Union two sets containing elements at indices a and b
        /// Returns true if a union was performed, false if already in same set
        pub fn unionSets(self: *Self, a: usize, b: usize) bool {
            const root_a = self.find(a);
            const root_b = self.find(b);

            if (root_a == root_b) {
                return false; // Already in same set
            }

            // Union by rank: attach smaller rank tree under root of higher rank tree
            if (self.nodes[root_a].rank < self.nodes[root_b].rank) {
                self.nodes[root_a].parent = root_b;
            } else if (self.nodes[root_a].rank > self.nodes[root_b].rank) {
                self.nodes[root_b].parent = root_a;
            } else {
                // Same rank: make one the root and increment its rank
                self.nodes[root_b].parent = root_a;
                self.nodes[root_a].rank += 1;
            }

            return true;
        }

        /// Get the value at an index
        pub fn getValue(self: Self, index: usize) T {
            std.debug.assert(index < self.count);
            return self.values[index];
        }

        /// Get the number of elements
        pub fn size(self: Self) usize {
            return self.count;
        }

        /// Count the number of distinct sets (connected components)
        pub fn countSets(self: *Self) usize {
            var set_count: usize = 0;
            for (0..self.count) |i| {
                if (self.find(i) == i) {
                    set_count += 1;
                }
            }
            return set_count;
        }

        /// Get all elements in the same set as the given element
        pub fn getSetMembers(self: *Self, allocator: Allocator, index: usize) Allocator.Error![]T {
            const root = self.find(index);
            var members = std.ArrayList(T).init(allocator);
            errdefer members.deinit();

            for (0..self.count) |i| {
                if (self.find(i) == root) {
                    try members.append(self.values[i]);
                }
            }

            return members.toOwnedSlice();
        }

        /// Get all sets as a list of lists
        pub fn getAllSets(self: *Self, allocator: Allocator) Allocator.Error![][]T {
            var sets_map = std.AutoHashMap(usize, std.ArrayList(T)).init(allocator);
            defer {
                var iter = sets_map.iterator();
                while (iter.next()) |entry| {
                    entry.value_ptr.deinit();
                }
                sets_map.deinit();
            }

            for (0..self.count) |i| {
                const root = self.find(i);
                const gop = try sets_map.getOrPut(root);
                if (!gop.found_existing) {
                    gop.value_ptr.* = std.ArrayList(T).init(allocator);
                }
                try gop.value_ptr.append(self.values[i]);
            }

            var result = std.ArrayList([]T).init(allocator);
            errdefer {
                for (result.items) |set| {
                    allocator.free(set);
                }
                result.deinit();
            }

            var iter = sets_map.iterator();
            while (iter.next()) |entry| {
                try result.append(try entry.value_ptr.toOwnedSlice());
            }

            return result.toOwnedSlice();
        }

        /// Get the size of the set containing the given element
        pub fn getSetSize(self: *Self, index: usize) usize {
            const root = self.find(index);
            var set_count: usize = 0;
            for (0..self.count) |i| {
                if (self.find(i) == root) {
                    set_count += 1;
                }
            }
            return set_count;
        }

        /// Reset all elements to individual sets
        pub fn reset(self: *Self) void {
            for (self.nodes, 0..) |*node, i| {
                node.parent = i;
                node.rank = 0;
            }
        }
    };
}

/// Simple DisjointSet for usize values (common use case)
pub const USizeDisjointSet = DisjointSet(usize);

// Tests
test "DisjointSet - basic union and find" {
    const allocator = std.testing.allocator;

    var ds = try DisjointSet(i32).init(allocator, &[_]i32{ 1, 2, 3, 4, 5 });
    defer ds.deinit();

    // Initially, each element is in its own set
    try std.testing.expectEqual(@as(usize, 5), ds.countSets());
    try std.testing.expect(ds.connected(0, 0));
    try std.testing.expect(!ds.connected(0, 1));

    // Union 0 and 1
    try std.testing.expect(ds.unionSets(0, 1));
    try std.testing.expect(ds.connected(0, 1));
    try std.testing.expectEqual(@as(usize, 4), ds.countSets());

    // Union 2 and 3
    try std.testing.expect(ds.unionSets(2, 3));
    try std.testing.expect(ds.connected(2, 3));
    try std.testing.expectEqual(@as(usize, 3), ds.countSets());

    // Union 0 and 2 (transitive)
    try std.testing.expect(ds.unionSets(1, 2));
    try std.testing.expect(ds.connected(0, 3));
    try std.testing.expectEqual(@as(usize, 2), ds.countSets());
}

test "DisjointSet - union already connected" {
    const allocator = std.testing.allocator;

    var ds = try DisjointSet(i32).init(allocator, &[_]i32{ 10, 20, 30 });
    defer ds.deinit();

    _ = ds.unionSets(0, 1);
    // Try to union again
    try std.testing.expect(!ds.unionSets(0, 1));
    try std.testing.expect(!ds.unionSets(1, 0));
}

test "DisjointSet - initRange" {
    const allocator = std.testing.allocator;

    var ds = try USizeDisjointSet.initRange(allocator, 10);
    defer ds.deinit();

    try std.testing.expectEqual(@as(usize, 10), ds.size());
    try std.testing.expectEqual(@as(usize, 10), ds.countSets());

    // Union even numbers
    _ = ds.unionSets(0, 2);
    _ = ds.unionSets(2, 4);
    _ = ds.unionSets(4, 6);
    _ = ds.unionSets(6, 8);

    try std.testing.expect(ds.connected(0, 8));
    try std.testing.expect(!ds.connected(0, 1));
    try std.testing.expectEqual(@as(usize, 6), ds.countSets());
}

test "DisjointSet - getSetMembers" {
    const allocator = std.testing.allocator;

    var ds = try DisjointSet(i32).init(allocator, &[_]i32{ 100, 200, 300, 400, 500 });
    defer ds.deinit();

    _ = ds.unionSets(0, 2);
    _ = ds.unionSets(1, 3);

    const set0 = try ds.getSetMembers(allocator, 0);
    defer allocator.free(set0);
    try std.testing.expectEqual(@as(usize, 2), set0.len);

    const set1 = try ds.getSetMembers(allocator, 1);
    defer allocator.free(set1);
    try std.testing.expectEqual(@as(usize, 2), set1.len);

    const set4 = try ds.getSetMembers(allocator, 4);
    defer allocator.free(set4);
    try std.testing.expectEqual(@as(usize, 1), set4.len);
}

test "DisjointSet - getSetSize" {
    const allocator = std.testing.allocator;

    var ds = try DisjointSet(i32).init(allocator, &[_]i32{ 1, 2, 3, 4, 5, 6 });
    defer ds.deinit();

    try std.testing.expectEqual(@as(usize, 1), ds.getSetSize(0));

    _ = ds.unionSets(0, 1);
    try std.testing.expectEqual(@as(usize, 2), ds.getSetSize(0));
    try std.testing.expectEqual(@as(usize, 2), ds.getSetSize(1));

    _ = ds.unionSets(2, 3);
    _ = ds.unionSets(0, 2);
    try std.testing.expectEqual(@as(usize, 4), ds.getSetSize(0));
}

test "DisjointSet - getAllSets" {
    const allocator = std.testing.allocator;

    var ds = try DisjointSet(i32).init(allocator, &[_]i32{ 1, 2, 3, 4, 5 });
    defer ds.deinit();

    _ = ds.unionSets(0, 1);
    _ = ds.unionSets(2, 3);

    const sets = try ds.getAllSets(allocator);
    defer {
        for (sets) |set| {
            allocator.free(set);
        }
        allocator.free(sets);
    }

    try std.testing.expectEqual(@as(usize, 3), sets.len);
}

test "DisjointSet - reset" {
    const allocator = std.testing.allocator;

    var ds = try DisjointSet(i32).init(allocator, &[_]i32{ 1, 2, 3, 4, 5 });
    defer ds.deinit();

    _ = ds.unionSets(0, 1);
    _ = ds.unionSets(2, 3);
    _ = ds.unionSets(0, 2);

    try std.testing.expectEqual(@as(usize, 2), ds.countSets());

    ds.reset();

    try std.testing.expectEqual(@as(usize, 5), ds.countSets());
    try std.testing.expect(!ds.connected(0, 1));
}

test "DisjointSet - path compression" {
    const allocator = std.testing.allocator;

    var ds = try USizeDisjointSet.initRange(allocator, 100);
    defer ds.deinit();

    // Create a long chain
    for (0..99) |i| {
        _ = ds.unionSets(i, i + 1);
    }

    // After find(0), all nodes should have shallow paths
    _ = ds.find(0);

    // Verify all are connected
    try std.testing.expect(ds.connected(0, 99));
}

test "DisjointSet - union by rank" {
    const allocator = std.testing.allocator;

    var ds = try USizeDisjointSet.initRange(allocator, 16);
    defer ds.deinit();

    // Build a balanced tree through union by rank
    _ = ds.unionSets(0, 1);
    _ = ds.unionSets(2, 3);
    _ = ds.unionSets(0, 2);
    _ = ds.unionSets(4, 5);
    _ = ds.unionSets(6, 7);
    _ = ds.unionSets(4, 6);
    _ = ds.unionSets(0, 4);

    // All should be connected
    try std.testing.expect(ds.connected(0, 7));
    try std.testing.expectEqual(@as(usize, 9), ds.countSets());
}

test "DisjointSet - getValue" {
    const allocator = std.testing.allocator;

    const values = [_]i32{ 10, 20, 30, 40, 50 };
    var ds = try DisjointSet(i32).init(allocator, &values);
    defer ds.deinit();

    try std.testing.expectEqual(@as(i32, 10), ds.getValue(0));
    try std.testing.expectEqual(@as(i32, 20), ds.getValue(1));
    try std.testing.expectEqual(@as(i32, 50), ds.getValue(4));
}

test "DisjointSet - single element" {
    const allocator = std.testing.allocator;

    var ds = try DisjointSet(i32).init(allocator, &[_]i32{42});
    defer ds.deinit();

    try std.testing.expectEqual(@as(usize, 1), ds.countSets());
    try std.testing.expect(ds.connected(0, 0));
    try std.testing.expect(!ds.unionSets(0, 0)); // Can't union with itself
}

test "DisjointSet - large set" {
    const allocator = std.testing.allocator;

    var ds = try USizeDisjointSet.initRange(allocator, 1000);
    defer ds.deinit();

    // Union in pairs
    var i: usize = 0;
    while (i < 1000) : (i += 2) {
        if (i + 1 < 1000) {
            _ = ds.unionSets(i, i + 1);
        }
    }

    try std.testing.expectEqual(@as(usize, 500), ds.countSets());
}

test "DisjointSet - findConst" {
    const allocator = std.testing.allocator;

    var ds = try USizeDisjointSet.initRange(allocator, 5);
    defer ds.deinit();

    _ = ds.unionSets(0, 1);
    _ = ds.unionSets(1, 2);

    // findConst should work without path compression
    const root = ds.findConst(2);
    try std.testing.expect(root < 3);
}

File: Python/disjoint_set_utils/disjoint_set.py
"""
Disjoint Set Union Implementation

Core implementation of Union-Find data structure with optimizations.
"""

from typing import Dict, List, Optional, Tuple, Generic, TypeVar, Set, Any
from abc import ABC, abstractmethod

T = TypeVar('T')


class DisjointSet(Generic[T]):
    """
    A Disjoint Set Union (Union-Find) data structure with path compression
    and union by rank optimizations.
    
    Time Complexity:
    - Find: O(α(n)) amortized, where α is the inverse Ackermann function
    - Union: O(α(n)) amortized
    - Space: O(n)
    
    Example:
        >>> ds = DisjointSet[int]()
        >>> ds.make_set(1)
        >>> ds.make_set(2)
        >>> ds.union(1, 2)
        >>> ds.connected(1, 2)
        True
        >>> ds.find(1) == ds.find(2)
        True
    """
    
    def __init__(self) -> None:
        """Initialize an empty disjoint set."""
        self._parent: Dict[T, T] = {}
        self._rank: Dict[T, int] = {}
        self._size: Dict[T, int] = {}  # Size of each component
        self._count: int = 0  # Number of disjoint sets
    
    def make_set(self, x: T) -> None:
        """
        Create a new set containing only element x.
        
        Args:
            x: The element to create a set for
            
        Raises:
            ValueError: If element already exists
        """
        if x in self._parent:
            raise ValueError(f"Element {x} already exists in the set")
        
        self._parent[x] = x
        self._rank[x] = 0
        self._size[x] = 1
        self._count += 1
    
    def make_sets(self, elements: List[T]) -> None:
        """
        Create multiple singleton sets.
        
        Args:
            elements: List of elements to create sets for
        """
        for x in elements:
            if x not in self._parent:
                self.make_set(x)
    
    def find(self, x: T) -> T:
        """
        Find the representative (root) of the set containing x.
        Uses path compression for optimization.
        
        Args:
            x: The element to find
            
        Returns:
            The representative element of the set containing x
            
        Raises:
            KeyError: If element x is not in any set
        """
        if x not in self._parent:
            raise KeyError(f"Element {x} not found in any set")
        
        # Path compression: make every node point directly to root
        if self._parent[x] != x:
            self._parent[x] = self.find(self._parent[x])
        
        return self._parent[x]
    
    def find_iterative(self, x: T) -> T:
        """
        Iterative version of find with two-pass path compression.
        
        Args:
            x: The element to find
            
        Returns:
            The representative element of the set containing x
        """
        if x not in self._parent:
            raise KeyError(f"Element {x} not found in any set")
        
        # Find root
        root = x
        while self._parent[root] != root:
            root = self._parent[root]
        
        # Path compression: point all nodes to root
        current = x
        while self._parent[current] != root:
            next_node = self._parent[current]
            self._parent[current] = root
            current = next_node
        
        return root
    
    def union(self, x: T, y: T) -> bool:
        """
        Unite the sets containing elements x and y.
        Uses union by rank for optimization.
        
        Args:
            x: First element
            y: Second element
            
        Returns:
            True if sets were united (were different), False if already in same set
        """
        root_x = self.find(x)
        root_y = self.find(y)
        
        if root_x == root_y:
            return False  # Already in the same set
        
        # Union by rank: attach smaller tree under root of larger tree
        if self._rank[root_x] < self._rank[root_y]:
            root_x, root_y = root_y, root_x
        
        self._parent[root_y] = root_x
        self._size[root_x] += self._size[root_y]
        
        # Increment rank if trees were same height
        if self._rank[root_x] == self._rank[root_y]:
            self._rank[root_x] += 1
        
        self._count -= 1
        return True
    
    def union_by_size(self, x: T, y: T) -> bool:
        """
        Unite sets using union by size strategy.
        
        Args:
            x: First element
            y: Second element
            
        Returns:
            True if sets were united, False if already in same set
        """
        root_x = self.find(x)
        root_y = self.find(y)
        
        if root_x == root_y:
            return False
        
        # Attach smaller tree under larger tree
        if self._size[root_x] < self._size[root_y]:
            root_x, root_y = root_y, root_x
        
        self._parent[root_y] = root_x
        self._size[root_x] += self._size[root_y]
        self._count -= 1
        
        return True
    
    def connected(self, x: T, y: T) -> bool:
        """
        Check if elements x and y are in the same set.
        
        Args:
            x: First element
            y: Second element
            
        Returns:
            True if in the same set, False otherwise
        """
        return self.find(x) == self.find(y)
    
    def get_size(self, x: T) -> int:
        """
        Get the size of the set containing element x.
        
        Args:
            x: The element
            
        Returns:
            Size of the component containing x
        """
        root = self.find(x)
        return self._size[root]
    
    def count_sets(self) -> int:
        """
        Get the number of disjoint sets.
        
        Returns:
            Number of disjoint sets
        """
        return self._count
    
    def get_sets(self) -> Dict[T, Set[T]]:
        """
        Get all disjoint sets as a dictionary mapping roots to their members.
        
        Returns:
            Dictionary mapping root elements to sets of members
        """
        sets: Dict[T, Set[T]] = {}
        for x in self._parent:
            root = self.find(x)
            if root not in sets:
                sets[root] = set()
            sets[root].add(x)
        return sets
    
    def get_members(self, x: T) -> Set[T]:
        """
        Get all members of the set containing element x.
        
        Args:
            x: The element
            
        Returns:
            Set of all elements in the same component as x
        """
        root = self.find(x)
        return {y for y in self._parent if self.find(y) == root}
    
    def __len__(self) -> int:
        """Return total number of elements in all sets."""
        return len(self._parent)
    
    def __contains__(self, x: T) -> bool:
        """Check if element x is in any set."""
        return x in self._parent
    
    def __repr__(self) -> str:
        return f"DisjointSet(sets={self._count}, elements={len(self._parent)})"


class WeightedDisjointSet(Generic[T]):
    """
    A Disjoint Set with support for weighted edges.
    Useful for problems involving relationships between elements.
    
    Example:
        >>> wds = WeightedDisjointSet[str]()
        >>> wds.make_set("a")
        >>> wds.make_set("b")
        >>> wds.union_weighted("a", "b", 5)
        >>> wds.same_component("a", "b")
        True
    """
    
    def __init__(self) -> None:
        """Initialize an empty weighted disjoint set."""
        self._parent: Dict[T, T] = {}
        self._weight: Dict[T, Dict[T, float]] = {}  # Edge weights
        self._rank: Dict[T, int] = {}
        self._component_weight: Dict[T, float] = {}  # Total weight per component
        self._count: int = 0
    
    def make_set(self, x: T) -> None:
        """Create a new singleton set."""
        if x in self._parent:
            raise ValueError(f"Element {x} already exists")
        
        self._parent[x] = x
        self._weight[x] = {}
        self._rank[x] = 0
        self._component_weight[x] = 0.0
        self._count += 1
    
    def make_sets(self, elements: List[T]) -> None:
        """
        Create multiple singleton sets.
        
        Args:
            elements: List of elements to create sets for
        """
        for x in elements:
            if x not in self._parent:
                self.make_set(x)
    
    def find(self, x: T) -> T:
        """Find the root of x with path compression."""
        if x not in self._parent:
            raise KeyError(f"Element {x} not found")
        
        if self._parent[x] != x:
            self._parent[x] = self.find(self._parent[x])
        
        return self._parent[x]
    
    def union_weighted(self, x: T, y: T, weight: float = 1.0) -> bool:
        """
        Unite sets with an associated weight.
        
        Args:
            x: First element
            y: Second element
            weight: Weight of the connection
            
        Returns:
            True if united, False if already connected
        """
        root_x = self.find(x)
        root_y = self.find(y)
        
        if root_x == root_y:
            return False
        
        if self._rank[root_x] < self._rank[root_y]:
            root_x, root_y = root_y, root_x
        
        self._parent[root_y] = root_x
        self._component_weight[root_x] += self._component_weight[root_y] + weight
        
        if self._rank[root_x] == self._rank[root_y]:
            self._rank[root_x] += 1
        
        self._count -= 1
        return True
    
    def get_component_weight(self, x: T) -> float:
        """Get total weight of the component containing x."""
        root = self.find(x)
        return self._component_weight[root]
    
    def same_component(self, x: T, y: T) -> bool:
        """Check if x and y are in the same component."""
        return self.find(x) == self.find(y)
    
    def count_sets(self) -> int:
        """Get the number of disjoint sets."""
        return self._count


class DisjointSetWithUndo(Generic[T]):
    """
    A Disjoint Set that supports undo operations.
    Uses persistent data structure techniques.
    
    Useful for backtracking algorithms and offline dynamic connectivity.
    """
    
    def __init__(self) -> None:
        """Initialize an empty undoable disjoint set."""
        self._parent: Dict[T, T] = {}
        self._rank: Dict[T, int] = {}
        self._size: Dict[T, int] = {}
        self._count: int = 0
        self._history: List[Tuple[str, T, T, int, int, int]] = []  # (op, a, b, rank_a, size_a, count)
    
    def make_set(self, x: T) -> None:
        """Create a new singleton set."""
        if x in self._parent:
            raise ValueError(f"Element {x} already exists")
        
        self._parent[x] = x
        self._rank[x] = 0
        self._size[x] = 1
        self._count += 1
        self._history.append(('make', x, x, 0, 1, self._count))
    
    def make_sets(self, elements: List[T]) -> None:
        """
        Create multiple singleton sets.
        
        Args:
            elements: List of elements to create sets for
        """
        for x in elements:
            if x not in self._parent:
                self.make_set(x)
    
    def find(self, x: T) -> T:
        """Find root without path compression (to support undo)."""
        if x not in self._parent:
            raise KeyError(f"Element {x} not found")
        
        while self._parent[x] != x:
            x = self._parent[x]
        
        return x
    
    def union(self, x: T, y: T) -> bool:
        """
        Unite sets containing x and y.
        Records operation for potential undo.
        """
        root_x = self.find(x)
        root_y = self.find(y)
        
        if root_x == root_y:
            self._history.append(('union_skip', x, y, 0, 0, self._count))
            return False
        
        # Record state before change
        old_rank_x = self._rank[root_x]
        old_size_x = self._size[root_x]
        old_count = self._count
        
        if self._rank[root_x] < self._rank[root_y]:
            root_x, root_y = root_y, root_x
        
        self._parent[root_y] = root_x
        self._size[root_x] += self._size[root_y]
        
        if self._rank[root_x] == self._rank[root_y]:
            self._rank[root_x] += 1
        
        self._count -= 1
        
        self._history.append(('union', root_y, root_x, old_rank_x, old_size_x, old_count))
        return True
    
    def undo(self) -> bool:
        """
        Undo the last operation.
        
        Returns:
            True if an operation was undone, False if no history
        """
        if not self._history:
            return False
        
        op, a, b, rank_a, size_a, count = self._history.pop()
        
        if op == 'make':
            del self._parent[a]
            del self._rank[a]
            del self._size[a]
            self._count = count
        elif op == 'union':
            # Restore parent, rank, and size
            self._parent[a] = a  # a was root_y, restore it as its own parent
            self._rank[b] = rank_a  # b was root_x
            self._size[b] = size_a
            self._count = count
        
        return True
    
    def connected(self, x: T, y: T) -> bool:
        """Check if x and y are in the same set."""
        return self.find(x) == self.find(y)
    
    def count_sets(self) -> int:
        """Get the number of disjoint sets."""
        return self._count
    
    def history_size(self) -> int:
        """Get the number of operations in history."""
        return len(self._history)


# ==================== Algorithm Helpers ====================

def count_connected_components(n: int, edges: List[Tuple[int, int]]) -> int:
    """
    Count the number of connected components in an undirected graph.
    
    Args:
        n: Number of vertices (0 to n-1)
        edges: List of edges as (u, v) tuples
        
    Returns:
        Number of connected components
        
    Example:
        >>> count_connected_components(5, [(0, 1), (1, 2), (3, 4)])
        2
    """
    ds = DisjointSet[int]()
    ds.make_sets(list(range(n)))
    
    for u, v in edges:
        ds.union(u, v)
    
    return ds.count_sets()


def detect_cycle_undirected(n: int, edges: List[Tuple[int, int]]) -> bool:
    """
    Detect if an undirected graph has a cycle.
    
    Args:
        n: Number of vertices (0 to n-1)
        edges: List of edges as (u, v) tuples
        
    Returns:
        True if cycle exists, False otherwise
        
    Example:
        >>> detect_cycle_undirected(3, [(0, 1), (1, 2), (2, 0)])
        True
        >>> detect_cycle_undirected(3, [(0, 1), (1, 2)])
        False
    """
    ds = DisjointSet[int]()
    ds.make_sets(list(range(n)))
    
    for u, v in edges:
        if ds.connected(u, v):
            return True  # Edge connects already connected vertices = cycle
        ds.union(u, v)
    
    return False


def kruskal_mst(n: int, edges: List[Tuple[int, int, float]]) -> Tuple[List[Tuple[int, int, float]], float]:
    """
    Find Minimum Spanning Tree using Kruskal's algorithm.
    
    Args:
        n: Number of vertices (0 to n-1)
        edges: List of edges as (u, v, weight) tuples
        
    Returns:
        Tuple of (MST edges, total weight)
        
    Example:
        >>> mst, weight = kruskal_mst(4, [(0, 1, 1), (1, 2, 2), (2, 3, 3), (0, 3, 4)])
        >>> weight
        6
    """
    # Sort edges by weight
    sorted_edges = sorted(edges, key=lambda e: e[2])
    
    ds = DisjointSet[int]()
    ds.make_sets(list(range(n)))
    
    mst_edges: List[Tuple[int, int, float]] = []
    total_weight = 0.0
    
    for u, v, w in sorted_edges:
        if ds.union(u, v):
            mst_edges.append((u, v, w))
            total_weight += w
            
            if len(mst_edges) == n - 1:
                break  # MST complete
    
    return mst_edges, total_weight


def find_redundant_connection(n: int, edges: List[Tuple[int, int]]) -> Optional[Tuple[int, int]]:
    """
    Find a redundant connection in a graph that would create a cycle.
    Returns the last edge that creates a cycle when added.
    
    Args:
        n: Number of vertices (1 to n)
        edges: List of edges as (u, v) tuples
        
    Returns:
        The redundant edge, or None if no redundancy
        
    Example:
        >>> find_redundant_connection(3, [(1, 2), (1, 3), (2, 3)])
        (2, 3)
    """
    ds = DisjointSet[int]()
    ds.make_sets(list(range(1, n + 1)))
    
    for u, v in edges:
        if ds.connected(u, v):
            return (u, v)
        ds.union(u, v)
    
    return None